boot/BootFromDevice.c

```diff
@@ -0,0 +1,364 @@
+
+	return result;
+}
+
+CONFIGENTRY* DetectSystemFatX(void) {
+	CONFIGENTRY* config = NULL;
+	CONFIGENTRY* cfgLinux;
+	FATXPartition* partition;
+
+	partition = OpenFATXPartition(0, SECTOR_STORE, STORE_SIZE);
+	if (partition == NULL) return NULL;
+
+	cfgLinux = DetectLinuxFATX(partition);
+	if (cfgLinux != NULL)
+	{
+		FillConfigEntries(cfgLinux, BOOT_FATX, 0, 0);
+		config = AddNestedConfigEntry(config, cfgLinux, "Linux");
+	}
+	CloseFATXPartition(partition);
+
+	return config;
+}
+
+int BootFromFatX(CONFIGENTRY *config) {
+	FATXPartition* partition;
+	int result = 0;
+
+	DVDTrayClose();
+
+	partition = OpenFATXPartition(0, SECTOR_STORE, STORE_SIZE);
+	if (partition == NULL) return false;
+
+	if (config->bootSystem == SYS_LINUX)
+		result = LoadLinuxFATX(partition, config);
+	CloseFATXPartition(partition);
+
+	return result;
+}
+
+CONFIGENTRY *DetectSystemCD(int cdromId) {
+	int n;
+	CONFIGENTRY* config = NULL;
+	CONFIGENTRY* cfgLinux;
+	int nTempCursorX, nTempCursorY;
+
+	printk("\2Please wait\n\n");
+	//See if we already have a CD in the drive
+	//Try for 8 seconds - takes a while to 'spin up'.
+	nTempCursorX = VIDEO_CURSOR_POSX;
+	nTempCursorY = VIDEO_CURSOR_POSY;
+again:
+	DVDTrayClose();
+	printk("Detecting system on CD... \n");
+	for (n = 0; n < 32; ++n) {
+		cfgLinux = DetectLinuxCD(cdromId);
+		if (cfgLinux != NULL) {
+			break;
+		}
+		wait_ms(250);
+	}
+
+	//We couldn't read the disk, so we eject the drive so the user can insert one.
+	if (cfgLinux == NULL) {
+		printk("Boot from CD failed.\nCheck that supported system exists on the CD.\n\n");
+		//Needs to be changed for non-xbox drives, which don't have an eject line
+		//Need to send ATA eject command.
+		DVDTrayEject();
+		wait_ms(2000); // Wait for DVD to become responsive to inject command
+
+		VIDEO_ATTR = 0xffeeeeff;
+		printk("\2Please insert CD and press Button A\n\n\2Press Button B to return to main menu\n\n");
+
+		while (1) {
+			// Retry system detection
+			cfgLinux = DetectLinuxCD(cdromId);
+
+			// Make button 'A' close the DVD tray
+			if (risefall_xpad_BUTTON(TRIGGER_XPAD_KEY_A) == 1) {
+				DVDTrayClose();
+				wait_ms(500);
+				break;
+			}
+			else if (DVD_TRAY_STATE == DVD_CLOSING) {
+				//It's an xbox drive, and somebody pushed the tray in manually
+				wait_ms(500);
+				break;
+			}
+			else if (cfgLinux != NULL) {
+				//It isnt an xbox drive, and somebody pushed the tray in manually, and
+				//the cd is valid.
+				break;
+			}
+			// Allow to cancel CD boot with button 'B'
+			else if (risefall_xpad_BUTTON(TRIGGER_XPAD_KEY_B) == 1) {
+				// Close DVD tray and return to main menu
+				DVDTrayClose();
+				wait_ms(500);
+				return NULL;
+			}
+			wait_ms(10);
+		}
+
+		wait_ms(250);
+
+		VIDEO_ATTR = 0xffffffff;
+
+		BootVideoClearScreen(&jpegBackdrop, nTempCursorY, VIDEO_CURSOR_POSY + 1);
+		VIDEO_CURSOR_POSX = nTempCursorX;
+		VIDEO_CURSOR_POSY = nTempCursorY;
+		goto again;
+	}
+
+	//Populate the configs with the drive ID
+	if (cfgLinux != NULL)
```

Comment:
We can reach here, if the system was not detected on CD, and user pressed button B to cancel boot from CD.

Reply:
I am new to C so forgive me if I'm missing something.

If a user hits the B button the the function short-circuits with a return `NULL`. In that case I also don't believe we'll hit this codepath at all.

Reply:
The function does not short-circuit, it returns immediately to the menu. Also I fixed the case of infinite loop (which happened before without user interaction on timeout) in the previous PR.

Reply:
I agree with @DiscoStarslayer : It can not be reached if `cfgLinux != NULL`, because there's a `goto again;` whenever it would have been `cfgLinux == NULL`.

Reply:
UPD: I read your comment wrong, and yeah you're right. But in later commits we will be able to reach this line with NULL value, as there will be another variable added.